beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java

```diff
@@ -657,6 +665,29 @@ public SafeFuture<Void> registerValidators(
             });
   }
 
+  @Override
+  public SafeFuture<Optional<DoppelgangerDetectionResult>> checkValidatorsDoppelganger(
+      List<UInt64> validatorIndices, final UInt64 epoch) {
+    return nodeDataProvider
+        .getValidatorLiveness(validatorIndices, epoch, chainDataProvider.getCurrentEpoch())
+        .thenApply(
+            validatorLivenessAtEpoches ->
+                Optional.of(
+                    getDoppelgangerDetectionResultFromLiveness(validatorLivenessAtEpoches)));
+  }
+
+  private DoppelgangerDetectionResult getDoppelgangerDetectionResultFromLiveness(
+      final Optional<List<ValidatorLivenessAtEpoch>> maybeValidatorsLiveness) {
+    Object2BooleanMap<UInt64> validatorsLiveness = new Object2BooleanArrayMap<>();
+    maybeValidatorsLiveness.ifPresent(
+        validatorLivenessAtEpoches ->
+            validatorLivenessAtEpoches.forEach(
+                validatorLivenessAtEpoch ->
+                    validatorsLiveness.put(
+                        validatorLivenessAtEpoch.getIndex(), validatorLivenessAtEpoch.isLive())));
+    return new DoppelgangerDetectionResult(validatorsLiveness);
```

Comment:
Why aren't we just returning the `VlaidatorsLiveness`? We want the return value for this method to be pretty much the same as what the REST API would return (we'll need to implement this method in `RemoteValidatorApi` by making a call to the liveness REST API).

It _can_ differ when there's a good reason to but given we have `NodeDataProvider.getValidatorLiveness` already returning reasonable internal types I'd have thought we'd just use those same types here.

Reply:
Encapsulating the result into `DoppelgangerDetectionResult` was an anticipation for the business logic implementation (Simplifying the result by removing the epoch since it will always be the same for all the validators and offering some helper methods...).
But we can definitely use the `List<ValidatorLivenessAtEpoch>>` returned by the `NodeDataProvider` without any encapsulation.
I'll change it to `List<ValidatorLivenessAtEpoch>>` and we'll see in the next step if such change is required.